### PR TITLE
CodeOwnersParser Team/User lookup

### PR DESCRIPTION
**Pull request by JimSuplizio**
When CodeOwnersParser encounters a team in the CODEOWNERS file it is parsing add the team's users, instead of the team, to the Owners list for the path. When adding the users for a team to the Owners list, it's done using a union to ensure that the resulting list is distinct. The list is pulled from blob data. 

In the case where a team is encountered in the CODEOWNERS file that doesn't exist in the team/user data, the team will get added as an owner which is the behavior today. Because of this, the ExcludeNonUserAliases function still needs to be in there.

I'd also fixed an [error](https://github.com/Azure/azure-sdk-tools/pull/6366/files#diff-f071416a991afecb5cbab353457ad32f1cbb8149ef4ba45db3c97e43a8b1befaR113) I'd encountered with parsing where, if a line didn't have any owner entries, the owner was being set to the path expression.  [Here's what the CODEOWNERS line looks like when this occurs.](https://github.com/Azure/azure-sdk-for-js/blob/main/.github/CODEOWNERS#L68)

The other piece I've added to the CodeOwnersParser solution is a manual test project, CodeOwnersManualTester, which loads the selected CODEOWNERS files for testing/inspection purposes. Since this is not the default project for the solution, debugging just requires right clicking on the project and selecting Debug from there.

### CODE REVIEW

File: tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs
Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs:13]: I'm OK with this be a default value but I think this needs to be an option that we can pass in at a command line parameter as well.

File: tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs:22]:
```suggestion
        public static Dictionary<string, List<string>>? GetTeamUserData()
```

Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs:22]:
```suggestion
        public static Dictionary<string, List<string>>? GetTeamUserData()
```

File: tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs:13]: Should `private static` fields start with small letter?

File: tools/code-owners-parser/CodeOwnersManualTester/CodeOwnerUtils.cs
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersManualTester/CodeOwnerUtils.cs:10]: I see inconsistency in naming:
- `CodeOwner`
- `CodeOwners`

I would suggest to always use `Codeowners`, as this is the convention I adopted in this codebase.

File: tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs
Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs:13]: Where is the tool/code that will create this file? Is that part of another PR?

File: tools/code-owners-parser/CodeOwnersManualTester/CodeOwnerUtils.cs
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersManualTester/CodeOwnerUtils.cs:18]: I feel this method belongs in the `CodeownersFile` class.

File: tools/code-owners-parser/CodeOwnersManualTester/CodeOwnersManualTester.csproj
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersManualTester/CodeOwnersManualTester.csproj:7]: Did you consider making them [errors](https://stackoverflow.com/a/62116924/986533) ?

File: tools/code-owners-parser/CodeOwnersManualTester/Program.cs
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersManualTester/Program.cs:5]: Do we need this tool in addition to [CodeownersManualAnalysisTests](https://github.com/Azure/azure-sdk-tools/blob/c37701ec43d68f96f480ddf7fc502841be7fbaed/tools/code-owners-parser/Azure.Sdk.Tools.RetrieveCodeOwners.Tests/CodeownersManualAnalysisTests.cs#L40) ?

File: tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs:131]: > In that case, just add the team to the list of authors

I understand this is our "best guess" what is going on, and this may actually be incorrect interpretation? If so, what bad things will happen?

File: tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs:134]: Should we also add `.Trim()` to `author` here?

File: tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs:113]: I think it would be nice to have automated unit tests covering this new logic that run as part of the automated regression test suite, not manually.

File: tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs:117]: To aid readability, I would extract the body of this `foreach` to a method with intention-revealing name.

File: tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs:22]: Will be updated with the next commit.

Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs:22]: Will be updated with next commit.

File: tools/code-owners-parser/CodeOwnersManualTester/CodeOwnerUtils.cs
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersManualTester/CodeOwnerUtils.cs:10]: Sure. I'll update with the next commit.

File: tools/code-owners-parser/CodeOwnersManualTester/CodeOwnerUtils.cs
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersManualTester/CodeOwnerUtils.cs:18]: The wrapper class is gone. Will be updated with the next commit.

File: tools/code-owners-parser/CodeOwnersManualTester/CodeOwnersManualTester.csproj
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersManualTester/CodeOwnersManualTester.csproj:7]: Sure, why not. Nothing in the ad-hoc project is using them but they'll be consistent with the rest of the projects in the solution.

File: tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs:13]: The tool that creates this is in the [GitHub Team User Store PR](https://github.com/Azure/azure-sdk-tools/pull/6355)

File: tools/code-owners-parser/CodeOwnersManualTester/CodeOwnersManualTester.csproj
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersManualTester/CodeOwnersManualTester.csproj:7]: I'm going to be removing this project altogether. You have something similar already in the solution.

File: tools/code-owners-parser/CodeOwnersManualTester/Program.cs
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersManualTester/Program.cs:5]: I'm removing this project with the next commit

File: tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs:134]: The fact that there was a trim for anything calling returned SplitLine call is a problem with that function. It's already using the `StringSplitOptions.RemoveEmptyEntries` in the call and should have been `StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries` to that line so everything returned is trimmed. Further, the documentation for [String.Split states](https://learn.microsoft.com/en-us/dotnet/api/system.stringsplitoptions?view=net-7.0) that if RemoveEmptyEntries and TrimEntries are specified together, then substrings that consist only of white-space characters are also removed from the result. Since string.split cannot return null, those IsNullOrWhiteSpace checks that are done on each entry can be removed. This will be updated with the next commit.

File: tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersParser/CodeownersEntry.cs:131]: This isn't a guess. The team is either in the team/user data and I'm adding the list of users for that team to the Owners instead of the team. If the team is not in the team/user data, then I'm just adding the team to the list as it previously did.

File: tools/code-owners-parser/Azure.Sdk.Tools.RetrieveCodeOwners/Program.cs
Comment (JimSuplizio) [inline tools/code-owners-parser/Azure.Sdk.Tools.RetrieveCodeOwners/Program.cs:55]: @weshaggard, I had to add another command line option to RetrieveCodeOwners to allow writing the json output to the file. The fact is, the [get-codeowners.lib.ps1](https://github.com/Azure/azure-sdk-tools/blob/952b6db75fd998adc6cf83af03da3a158ab1ee80/eng/common/scripts/get-codeowners.lib.ps1#L125) was incorrectly assuming any/all output from the executable was nothing more than a json codeowner entry. Using CodeOwnersParser FileHelper [GetUrlContents](https://github.com/Azure/azure-sdk-tools/blob/main/tools/code-owners-parser/CodeOwnersParser/FileHelpers.cs#L38) to fetch the blob, will cause output that'll break this assumption. The correct fix for this is the add the option to write the entry to a file. The problem here is because of locked version in the [eng/common/get-codeowners.lib.ps1](https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/scripts/get-codeowners.lib.ps1#L81) script, I can't make the changes, to use the file instead of output, to the script until this is checked in and another version has been published.
@konrad-jamrozik and I already chatted about this and both agreed this was the correct way to fix this.

File: tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs]: As per our discussion, this is now an option that can be given to the parser and has been added as a command line option to the RetrieveCodeOwners executable.

File: tools/code-owners-parser/CodeOwnersParser/TeamUserHolder.cs
Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/TeamUserHolder.cs:50]: Probably worth including the TeamUserStorageURI as part of the error message as well.

File: tools/code-owners-parser/CodeOwnersParser/TeamUserHolder.cs
Comment (weshaggard) [inline tools/code-owners-parser/CodeOwnersParser/TeamUserHolder.cs:71]: We might want to think about making this logging configurable or use a logging interface at some point as generally it isn't great for shared libraries to have a bunch of logging.

**Review by weshaggard — APPROVED** [review]

File: tools/code-owners-parser/CodeOwnersParser/TeamUserHolder.cs
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersParser/TeamUserHolder.cs:50]: I'll add this in another commit before I merge.

File: tools/code-owners-parser/CodeOwnersParser/TeamUserHolder.cs
Comment (JimSuplizio) [reply tools/code-owners-parser/CodeOwnersParser/TeamUserHolder.cs:71]: That won't be part of this PR. I created a [new issue](https://github.com/Azure/azure-sdk-tools/issues/6438) to track this.

**Pull request merged**
